pkg/executor/detach.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"github.com/pingcap/tidb/pkg/executor/internal/exec"
+	"github.com/pingcap/tidb/pkg/expression/contextsession"
+)
+
+// Detach detaches the current executor from the session context. After detaching, the session context
+// can be used to execute another statement while this executor is still running. The returning value
+// shows whether this executor is able to be detached.
+//
+// NOTE: the implementation of `Detach` should guarantee that no matter whether it returns true or false,
+// both the original executor and the returning executor should be able to be used correctly. This restriction
+// is to make sure that if `Detach(a)` returns `true`, while other children of `a`'s parent returns `false`,
+// the caller can still use the original one.
+func Detach(originalExecutor exec.Executor) (exec.Executor, bool) {
+	newExecutor, ok := originalExecutor.Detach()
+	if !ok {
+		return nil, false
+	}
+
+	children := originalExecutor.AllChildren()
+	newChildren := make([]exec.Executor, len(children))
+	for i, child := range children {
+		detached, ok := Detach(child)
+		if !ok {
+			return nil, false
+		}
+		newChildren[i] = detached
+	}
+	copy(newExecutor.AllChildren(), newChildren)
+
+	return newExecutor, true
+}
+
+func (treCtx tableReaderExecutorContext) Detach() tableReaderExecutorContext {
+	newCtx := treCtx
+
+	if ctx, ok := treCtx.ectx.(*contextsession.SessionExprContext); ok {
+		staticExprCtx := ctx.IntoStatic()
```

Comment:
Is it possible to make `IntoStatic` a static function and receives `exprctx.BuildContext` as input? So that we will need to try to cast the expr ctx to `SessionExprContext`

Reply:
It's a little weird. I think it'd be better to announce that it can only handle the `SessionExprContext`, as we may have many other implementations for `exprctx.BuildContext`.

For example, the `NullRejectCheckExprContext { *SessionExprContext }` also implements the `exprctx.BuildContext`. If we are going to implement a standalone static `IntoStatic` function, it'll have to handle this case. It's not easy to maintain all possible implementations in that function (though the only possible branch is the `*contextsession.SessionExprContext`).